### DISCUSSION TITLE
Support additional styles for the List widget

### DISCUSSION THREAD

**Discussion by orhun** [ratatui/ratatui#1168] [Ideas]
I really like the styling in the [`bpytop`](https://github.com/aristocratos/bpytop)'s list widget (written in Python):

![rec_20240604T232316](https://github.com/ratatui-org/ratatui/assets/24392180/a43eeedb-a8af-45e2-9974-28f7b0494dc6)

Supporting something like that (fading as you scroll down) in Ratatui (preferably as "out of the box") would be really nice I think.

Just an idea, I'm dropping this here to get opinions & talk about possible ways of implementing it.

**Comment by joshka** [comment]
Nice

I was thinking a bit about rendering rows in tables / lists recently on a similar note. Right now we pass in to a list, an Iterator of Into<ListItem>, but do the conversion immediately. Really we should be deferring the render until the actual lines are needed. This is sort of the same idea, because instead of style being a constant value, we're thinking about this as a value calculated per ListItem.

The same idea could also be more generally relevant to Tables, Tabs, the various gauges and charts. In effect, replace constants with calculations.

**Reply by orhun** [reply]
Yes, doing this requires calculating the style on the fly based on the position in the list so that approach definitely makes sense. However, I didn't fully understand if that will be an internal change or user-facing one?

I was thinking of supporting this with no additional code to the user's app except a new property such as `.scroll_style(ScrollStyle::Fading)` or `.fading_scroll(true)`.

**Reply by joshka** [reply]
Adding this as a configuration option seems like it's too specific to this particular display to me, and it would suggest the following questions:
- What sort of things would need to be configured?
- What would the default values for colors / fading amounts be?
- How many lines should be faded?
- How is the up / down movement handled?

I definitely think this is a good idea, but I'd look at it from a perspective of taking a step back and working out how to make something like this configurable by additive code instead of modifying the existing behavior (i.e. inject styles based on the state / line values instead of change the rendering based on a configuration value). (See https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle for this in OOP langugages - the principle sort of still applies here to rust though)

**Reply by orhun** [reply]
I see your point and just got a flashback to the Java classes in college. Dark times.

So what do you think is the best step towards making this possible? I can work on a custom widget and try to hack it together, but I think it is better to have an extension point first, where we are able to _calculate_ the items to render.

**Reply by joshka** [reply]
I'd experiment a bit with what it takes to make each line that's rendered have some different rendering. I.e. insert a closure / trait implementation / ? that takes some info on the line / item being rendered and returns either style info or performs the rendering. Effectively this approach is a "replace concrete behavior with strategy pattern" refactoring. The devil on this is in the details of configuring this, where to be generic vs concrete, working out what params need to be passed to the strategy (List/ListItem/State), and then actually making it work.

**Comment by orhun** [comment]
I implemented the fading list in [3b69cfebc3c3afac52f223acc929013a6b614e3b](https://github.com/orhun/theattyr/commit/3b69cfebc3c3afac52f223acc929013a6b614e3b) and it was quite straightforward:

```rust
let mut lines = Vec::new();
let selected = self.list_state.selected().unwrap_or_default();
for (i, animation) in self.animations.iter().enumerate() {
    let distance = (i as isize - selected as isize).abs() as f64;
    let adjusted_brightness = (100. - distance * 2.).max(30.);
    let color = Color::from_hsl(0.0, 0.0, adjusted_brightness);
    let line = animation.to_line().style(Style::default().fg(color));
    lines.push(line);
}
```

**Reply by joshka** [reply]
Yeah, this is the sort of thing I'd expect. Because there's a bunch of different things you could do with styling this, it's likely better to do it outside the widget and pass the items in like this.

**Reply by orhun** [reply]
Yup, I will close this for now, and we can revisit it if there is interest in the future.

![rec_20240926T234010](https://github.com/user-attachments/assets/fbd4e1c6-b5a7-49b1-8d4e-516341ac7d25)

**Reply by joshka** [reply]
that's pretty

**Reply by joshka** [reply]
This could make a good recipe for the website perhaps

**Reply by orhun** [reply]
https://github.com/ratatui/ratatui-website/issues/736
